apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java

```diff
@@ -134,8 +137,26 @@ private void onTransactionStart(boolean startedAsChild, long epochMicros, Sample
         spanCompressionEnabled = spanConfig.isSpanCompressionEnabled();
         spanCompressionExactMatchMaxDurationUs = spanConfig.getSpanCompressionExactMatchMaxDuration().getMicros();
         spanCompressionSameKindMaxDurationUs = spanConfig.getSpanCompressionSameKindMaxDuration().getMicros();
+
         if (!startedAsChild) {
             traceContext.asRootSpan(sampler);
+        } else {
+            CoreConfiguration.TraceContinuationStrategy traceContinuationStrategy =
+                tracer.getConfig(CoreConfiguration.class).getTraceContinuationStrategy();
+
+            if (traceContinuationStrategy.equals(RESTART) ||
+                (
+                    traceContinuationStrategy.equals(RESTART_EXTERNAL) &&
+                        !traceContext.getTraceState().includesVendor()
+                )
+            ) {
+                //Add a span link to the parent
+                addSpanLink(TraceContext.fromParentContext(), traceContext.copy().setIdToParentId());
```

Comment:
Why is `setIdToParentId()` needed? Isn't the id of a span link irrelevant? We're only serializing the trace id and span id of span links. See also `co.elastic.apm.agent.report.serialize.DslJsonSerializer#serializeSpanLinks`.

```suggestion
                addSpanLink(TraceContext.fromParentContext(), traceContext.copy());
```

Reply:
You are right the ID of the span link is irrelevant but we need the call because ... (this is part of the add span links implementation that Eyal wasn't that happy with). The span link added is the CHILD of the trace context passed here, so the resulting span link's parent ID is the ID of the tracecontext, not the parent ID. ie the general case is
  `TraceContext(traceID=A, ID=B, ParentID=C) -> spanlink(traceID=A, ID=D, ParentID=B)`

Reply:
But for a span link, the span.id is irrelevant, right? When calling `setIdToParentId()`, we'd have `spanlink(traceID=A, ID=B, ParentID=B)` i.e. the ID and the ParentID being the same value which also doesn't seem quite right.

Reply:
Yes. But the span link parent needs to point to the trace context parent. So let me work through it

1. The external span is "X-OtherServiceSpan". We want the span link to point to that
2. First we construct a TraceContext with `TraceID=X, ID=NewSpanID-A, ParentID=OtherServiceSpan`
3. If we construct a span link from this TraceContext as is, we would have a SpanLink with `TraceID=X, ID=NewSpanID-B, ParentID=NewSpanID-A` - because the span link creates a child of that TraceContext. If you send this to the APM server, you send `SpanLink(TraceID=X, ParentID=NewSpanID-A)` - the span link is pointing at a span which doesn't exist anywhere in the APM server. We want the span link to point at the incoming request, which is `SpanLink(TraceID=X, ParentID=OtherServiceSpan)`
4. So in order to achieve this, I overwrite the ID, ie we start with `TraceContext(TraceID=X, ID=NewSpanID-A, ParentID=OtherServiceSpan)`, then I overwrite the ID to get `TraceContext(TraceID=X, ID=OtherServiceSpan, ParentID=OtherServiceSpan)`, then I construct the span link (which is a child of that) to get `SpanLink(TraceID=X, ID=DontCare, ParentID=OtherServiceSpan)`.
5. Then this `SpanLink(TraceID=X, ID=DontCare, ParentID=OtherServiceSpan)` is serialized (just `TraceID=X, ParentID=OtherServiceSpan` ie a link pointing at "X-OtherServiceSpan") and the APM server has the span link we want, pointing at that incoming span

Reply:
Ah, got it now. The part that I was missing is that `addSpanLink` creates a child context of the provided context. Since you don't have the original context that has been provided in the headers of the incoming request, you re-create a "fake" parent so that you can create another child.

It's a few more hoops than I would have expected.

In a nutshell, what we currently have is:
- Make transaction a child of the remote parent
- If the trace should be re-started
  - Modify the context so that the transaction is a new trace root
  - Re-create the remote parent context
  - Add a span link, creating a child context of the fake remote context

I think we could simplify by moving the logic from `onTransactionStart` to `start`:

- If the trace should be re-started
  - Add a span link, creating a child context of the remote context
- Else
  - Make transaction a child of the remote parent

Reply:
I started with start(), but it has an implementation for binary and text, which means doing it there the code is partially duplicated, and to create the span link there I have to handle the separate implementations. I'm not sure it simplifies the code. I'll check

Reply:
Right, the `start` method is overloaded. But it's not binary vs text based. The first one is used only to create a root transaction and the second `start` method is only used when there's a remote context. I'd suggest renaming the methods to make that more clear. The span link thing is only relevant for the second `start` method.

Reply:
The start -> startRoot for the first is much better, thanks! Otherwise, I've reimplemented per your suggestion, I don't actually think it's clearer. But it's not any worse either

Reply:
I made some suggestions to make the implementation a bit cleaner by avoiding instanceof checks and casts: https://github.com/jackshirazi/apm-agent-java/pull/3

Reply:
Very nice, much cleaner like that!